source/common/grpc/async_client_manager_impl.cc

```diff
@@ -138,5 +141,38 @@ AsyncClientManagerImpl::factoryForGrpcService(const envoy::config::core::v3::Grp
   return nullptr;
 }
 
+RawAsyncClientSharedPtr AsyncClientManagerImpl::getOrCreateRawAsyncClient(
+    const envoy::config::core::v3::GrpcService& config, Stats::Scope& scope,
+    bool skip_cluster_check, bool always_use_cache) {
+  if (!always_use_cache &&
+      !Runtime::runtimeFeatureEnabled("envoy.reloadable_features.enable_grpc_async_client_cache")) {
+    return factoryForGrpcService(config, scope, skip_cluster_check)->createUncachedRawAsyncClient();
+  }
+  RawAsyncClientSharedPtr client;
+  client = raw_async_client_cache_->getCache(config);
+  if (client != nullptr) {
+    return client;
+  }
+  client = factoryForGrpcService(config, scope, skip_cluster_check)->createUncachedRawAsyncClient();
+  raw_async_client_cache_->setCache(config, client);
+  return client;
+}
+
+RawAsyncClientSharedPtr AsyncClientManagerImpl::RawAsyncClientCache::getCache(
+    const envoy::config::core::v3::GrpcService& config) const {
+  uint64_t key = MessageUtil::hash(config);
+  auto it = cache_.find(key);
```

Comment:
I'm still feeling paranoid that the behavior on collision may be poor (even if rare). Wouldn't it be easy to use the hash as a hash, and use message-comparison as the comparator?

Reply:
but currently we don't store the (const envoy::config::core::v3::GrpcService& config) inside the client, do you think we should store it as a member for hash comparison?

Reply:
Yup that's a good solution. Or you could just use the GrpcService as the key to a map. It looks pretty small.